src/coreclr/jit/importercalls.cpp

```diff
@@ -3987,6 +3988,49 @@ GenTree* Compiler::impSRCSUnsafeIntrinsic(NamedIntrinsic        intrinsic,
         {
             assert((sig->sigInst.methInstCount == 1) || (sig->sigInst.methInstCount == 2));
 
+            // The signature is:
+            //
+            //   T As<T>(object o) where T : class
+            //
+            // it means we need to query the exact T since it's always a shared type due
+            // to its constraint.
+            //
+            if (sig->sigInst.methInstCount == 1)
+            {
+                CORINFO_GENERICHANDLE_RESULT embedInfo;
+                info.compCompHnd->embedGenericHandle(pResolvedToken, false, &embedInfo);
```

Comment:
`embedGenericHandle` should be only used when you actually want to burn the handle into the code. It should not be used for type system queries to support optimizations.